prysmaticlabs/prysm
# Attempt to reconnect when waiting for activation

**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

When the connection between the validator and beacon chain is interrupted while waiting for any key to be activated, the validator fatally crashes.
This can be a bit of an issue when validators are waiting days or weeks to be activated and the operator attempts to restart or update their beacon node(s).

**Which issues(s) does this PR fix?**

Fixes #2243

**Other notes for review**

I moved code related to WaitForActivation from validator.go to its own go file.

Tested briefly locally in pyrmont.

Testing steps:
- Sync beacon node
- Start a validator with `n` never-activated keys.
- Observe log that validator is waiting for activation
- Stop the beacon node
- Observe error logs from validator
- Restart beacon node
- Observe log that validator is waiting for activation


Example logs:
```
[2020-12-06 11:08:56]  INFO validator: Waiting for beacon node to sync to latest chain head
[2020-12-06 11:09:02]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0x99e3bb36f3b2 status=UNKNOWN_STATUS
[2020-12-06 11:09:02]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0xa40c681aadca status=UNKNOWN_STATUS
[2020-12-06 11:09:02]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0xa73c75b4f28d status=UNKNOWN_STATUS
[2020-12-06 11:09:02]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0x9450379ab58e status=UNKNOWN_STATUS
[2020-12-06 11:09:02]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0xb058bdf768a0 status=UNKNOWN_STATUS
[2020-12-06 11:09:04] ERROR validator: Stream broken while waiting for activation. Reconnecting... attempts=1 error=rpc error: code = Canceled desc = RPC context canceled
[2020-12-06 11:09:09] ERROR validator: Stream broken while waiting for activation. Reconnecting... attempts=2 error=rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:4000: connect: connection refused"
[2020-12-06 11:09:15] ERROR validator: Stream broken while waiting for activation. Reconnecting... attempts=3 error=rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:4000: connect: connection refused"
[2020-12-06 11:09:22] ERROR validator: Stream broken while waiting for activation. Reconnecting... attempts=4 error=rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:4000: connect: connection refused"
[2020-12-06 11:09:30] ERROR validator: Stream broken while waiting for activation. Reconnecting... attempts=5 error=rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:4000: connect: connection refused"
[2020-12-06 11:09:39] ERROR validator: Stream broken while waiting for activation. Reconnecting... attempts=6 error=rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:4000: connect: connection refused"
[2020-12-06 11:09:48]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0x99e3bb36f3b2 status=UNKNOWN_STATUS
[2020-12-06 11:09:48]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0xa40c681aadca status=UNKNOWN_STATUS
[2020-12-06 11:09:48]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0xa73c75b4f28d status=UNKNOWN_STATUS
[2020-12-06 11:09:48]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0x9450379ab58e status=UNKNOWN_STATUS
[2020-12-06 11:09:48]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0xb058bdf768a0 status=UNKNOWN_STATUS
```
